File: largest-color-value-in-a-directed-graph.cc
// Largest Color Value in a Directed Graph
#define FOR(i, a, b) for (remove_cv<remove_reference<decltype(b)>::type>::type i = (a); i < (b); i++)
#define REP(i, n) FOR(i, 0, n)

class Solution {
public:
  int largestPathValue(string colors, vector<vector<int>>& edges) {
    int n = colors.size(), top = -1, cnt = 0;
    vector<vector<int>> es(n);
    int ans = 0;
    for (auto &x: edges)
      es[x[0]].push_back(x[1]);

    REP(cc, 26) {
      vector<int> d(n), c(n);
      for (auto &x: edges)
        d[x[1]]++;
      top = -1;
      REP(i, n) {
        if (!d[i]) {
          d[i] = top;
          top = i;
        }
      }
      while (~top) {
        int u = top;
        top = d[top];
        cnt++;
        c[u] += colors[u] == 'a'+cc;
        for (int v: es[u]) {
          c[v] = max(c[v], c[u]);
          if (!--d[v]) {
            d[v] = top;
            top = v;
          }
        }
        ans = max(ans, c[u]);
      }
      if (cnt < n) return -1;
    }
    return ans;
  }
};

File: largest-color-value-in-a-directed-graph.rs
impl Solution {
    pub fn largest_path_value(colors: String, edges: Vec<Vec<i32>>) -> i32 {
        let n = colors.len();
        let mut deg = vec![0; n];
        let mut head = vec![0; n];
        let mut es: Vec<(usize, usize)> = vec![(0,0)];
        for e in edges.iter() {
            deg[e[1] as usize] += 1;
            es.push((e[1] as usize, head[e[0] as usize]));
            head[e[0] as usize] = es.len()-1;
        }
        let mut q = vec![];
        for i in 0..n {
            if deg[i] == 0 {
                q.push(i); }}
        let mut i = 0;
        while i < q.len() {
            let u = q[i];
            i += 1;
            let mut j = head[u];
            while j != 0 {
                let v = es[j].0;
                j = es[j].1;
                deg[v] -= 1;
                if deg[v] == 0 { q.push(v); }
            }
        }
        if q.len() != n { return -1; }

        let mut ans = 0;
        for c in b'a'..(b'z'+1) {
            let mut dp = vec![0; n];
            for &u in q.iter() {
                if colors.as_bytes()[u] == c { dp[u] += 1; }
                let mut j = head[u];
                while j != 0 {
                    let v = es[j].0;
                    j = es[j].1;
                    dp[v] = dp[v].max(dp[u]);
                }
            }
            ans = dp.iter().fold(ans, |acc,&v| acc.max(v));
        }
        ans
    }
}
